src/rust/lib/graph-editor/src/component/node/port/output.rs

```diff
@@ -0,0 +1,300 @@
+            let pos   = Vector2::new(pos_x,pos_y);
+            view.set_position_xy(pos);
+
+            let shape = &view.shape;
+            shape.sprite.size().set(element_size);
+            shape.shape_width.set(width);
+            shape.padding.set(self.gap_width);
+            shape.offset_x.set(x_delta * index as f32);
+        }
+    }
+
+    fn update_shapes(&mut self) {
+        self.update_ports();
+    }
+
+    fn set_size(&mut self, size:Vector2<f32>) {
+        self.size = size;
+        self.update_shapes();
+    }
+}
+
+// ===================
+// === OutPutPorts ===
+// ===================
+
+#[derive(Debug,Clone,CloneRef)]
+pub struct OutputPorts {
+    pub frp            : Frp,
+        network        : frp::Network,
+        data           : Rc<RefCell<OutPutPortsData>>,
+        display_object : display::object::Instance,
+}
+
+impl OutputPorts {
+    pub fn new(scene:&Scene, number_of_ports:u8) -> Self {
+        let logger         = Logger::new("bubble");
+
+        let display_object = display::object::Instance::new(&logger);
+        let network        = default();
+        let frp            = Frp::new(&network);
+
+        let size           = Vector2::zero();
+        let scene          = scene.clone_ref();
+        let gap_width      = SEGMENT_GAP_WIDTH;
+
+        let mut ports      = Vec::default();
+        ports.resize_with(number_of_ports as usize,|| component::ShapeView::new(&logger,&scene));
+        let ports = ports;
+
+        let data =  OutPutPortsData {display_object:display_object.clone_ref(),scene,logger,size,
+                                     ports,gap_width}.init();
+        let data = Rc::new(RefCell::new(data));
+
+        OutputPorts {data,network,frp,display_object}.init()
+    }
+
+    fn init(mut self) -> Self {
+        self.init_frp();
+        self
+    }
+
+    fn init_frp(&mut self) {
+        let network = &self.network;
+        let frp     = &self.frp;
+        let data    = &self.data;
+
+        frp::extend! { network
+            eval  frp.set_size ((size) data.borrow_mut().set_size(size.unwrap_or_else(zero)));
+
+            def set_port_size      = source::<(PortId,f32)>();
+            def set_port_sizes_all = source::<f32>();
+
+            def _set_port_sizes = set_port_sizes_all.map(f!([set_port_size,data](size) {
+                let port_num = data.borrow().ports.len();
+                for index in 0..port_num {
+                    set_port_size.emit((index,*size));
+                };
+            }));
+
+            def set_port_opacity     = source::<(PortId,f32)>();
+            def set_port_opacity_all = source::<f32>();
+
+            def _set_port_opacities = set_port_opacity_all.map(f!([set_port_opacity,data](size) {
+                let port_num = data.borrow().ports.len();
+                for index in 0..port_num {
+                    set_port_opacity.emit((index,*size));
+                };
+            }));
+        }
+
+        // Init ports
+        {
```

Comment:
create a separate function for that instead of such brackets. This is hacky and its hard to understand whats happening here. Moreover, you are borrowing value in the next line and it is borrowed for the next 50 lines. It is really hard to track down. You are in the initialization stage. This should be possible to be expressed without any borrow.